libs/prelude/Prelude/Interfaces.idr

```diff
@@ -234,6 +234,11 @@ public export
 concatMap : (Foldable t, Monoid m) => (a -> m) -> t a -> m
 concatMap f = foldr ((<+>) . f) neutral
 
+|||Test whether the structure is empty.
+public export
+null : Foldable t => t a -> Bool
+null = foldr (\_,_ => False) True
```

Comment:
Because Idris is not lazy, it's probably a bad idea to write it this way:
`foldr c n (x :: xs) = c n (foldr c n xs)` means that the recursive call
will be fully evaluated before `c` is called.

This means that `null` is linear in the size of the `t a` input.
It's probably better to add it to the `Foldable` class and let users
implement it efficiently for each data type.

Reply:
I am sorry for not being considerate and thanks for your patience.

Reply:
I did not say that the PR would not get accepted. I just said `null` should
be moved to the definition of `interface Foldable` and then each implementation
should provide its own efficient instance of `null`.

This is a welcomed addition!

Reply:
Wow! Thank You. I had done some fix about them …… testing XD

Reply:
![图片](https://user-images.githubusercontent.com/68433824/101662513-9122c080-3a84-11eb-964b-7dac51d4faff.png)
![图片](https://user-images.githubusercontent.com/68433824/101662862-f7a7de80-3a84-11eb-87c2-a6cafcd3813d.png)
The following code can be found at this [site](http://fars.ee/h8Qo).
Sorry to bother. What I had just done is to
- Add Foldable implementations for `List`, `Maybe` and `Either`, respectively.
- Add `null` in the `Foldable` interface.
- Add a default definition for `null` for compatibility and a flag `%allow_overloads null`.
I am wondering how to write the code correspondence to `Foldable Either`. As presentations in the pics, I got confusing.
Feeling sorry...
:(

Reply:
Maybe the Evaluation strategy?
Since the type of
`Left 42` : Either Integer ?b
`Right 42` : Either ?a Integer

Reply:
Note that it's a convention to see `Either a` (given `a` arbitrary type variable) as a functor, but we can also see that as a consequence of GHC's lack of type lambda support. If we only see `Either` as a bifunctor this won't be a problem but there is another convention, that is to use `Either` as the error monad, and that requires the `Either a` functor (as well as the `Either a` foldable).